### PR TITLE
src/encoding/proto: Refactor EncodeLabel and EncodeMetric

### PR DESCRIPTION
**Pull request by mxinden**
This patch would remove most of the `collect` calls. @ackintosh what do you think? Would you mind running this through your allocation counter?

### CODE REVIEW

**Comment by mxinden** [comment]
My changes are not well documented. Happy to answer any questions here, or do a small write-up.

**Comment by ackintosh** [comment]
I run this with the allocation counter.  It was `47`, the count was reduced. 🎉  ([49](https://github.com/ackintosh/client_rust/pull/2#issuecomment-1203634863) → 47)


```shell
$ cargo test --lib encoding::proto::tests::encode_counter_family_with_prefix_with_label --all-features -- --show-output

successes:

---- encoding::proto::tests::encode_counter_family_with_prefix_with_label stdout ----
allocation: 47
```

File: src/encoding/proto.rs [hunk 1]
```diff
@@ -359,14 +358,17 @@ where
...
 impl EncodeMetric for Histogram {
-    type Iterator = IntoIter<openmetrics_data_model::Metric>;
+    type Iterator = std::iter::Once<openmetrics_data_model::Metric>;
 
     fn encode(&self, labels: Vec<openmetrics_data_model::Label>) -> Self::Iterator {
         let (sum, count, buckets) = self.get();
         // TODO: Would be better to use never type instead of `()`.
-        let mut metric = encode_histogram_with_maybe_exemplars::<()>(sum, count, &buckets, None);
+        // TODO: Revert (String, String)?
+        let mut metric = encode_histogram_with_maybe_exemplars::<Vec<(String, String)>>(
```
Comment (ackintosh) [inline src/encoding/proto.rs:367]: > // TODO: Revert (String, String)?

I think this is good, no need to revert. 💡

File: src/encoding/proto.rs [hunk 2]
```diff
@@ -117,59 +121,49 @@ impl EncodeMetric
     }
 }
 
+// TODO: Rename to EncodeLabels?
```
Comment (ackintosh) [inline src/encoding/proto.rs:124]: Agree with that as `encode()` returns one or more labels.

File: src/encoding/proto.rs [hunk 3]
```diff
@@ -329,25 +323,30 @@ where
 /////////////////////////////////////////////////////////////////////////////////
 // Family
 
-impl<S, M, C> EncodeMetric for Family<S, M, C>
+impl<'c, S, M, C> EncodeMetric for Family<S, M, C>
```
Comment (ackintosh) [inline src/encoding/proto.rs:326]: Is the lifetime parameter `'c` necessary?

**Comment by ackintosh** [comment]
Note for self:  [Higher-Rank Trait Bounds - The Rustonomicon](https://doc.rust-lang.org/nomicon/hrtb.html)

I learned a lot, thank you!

**Comment by mxinden** [comment]
> I run this with the allocation counter. It was `47`, the count was reduced. tada ([49](https://github.com/ackintosh/client_rust/pull/2#issuecomment-1203634863) → 47)
> 
> ```shell
> $ cargo test --lib encoding::proto::tests::encode_counter_family_with_prefix_with_label --all-features -- --show-output
> 
> successes:
> 
> ---- encoding::proto::tests::encode_counter_family_with_prefix_with_label stdout ----
> allocation: 47
> ```

I think the actual benchmark comparison would be more fruitful. E.g. running on both versions and then comparing via https://github.com/BurntSushi/critcmp.

File: src/encoding/proto.rs (hunk 3 above)
Comment (mxinden) [reply src/encoding/proto.rs:326]: Oh, I probably missed removing it after experimenting.

File: src/encoding/proto.rs (hunk 1 above)
Comment (mxinden) [reply src/encoding/proto.rs:367]: We could as well use https://docs.rs/void/1.0.2/void/enum.Void.html, thus making this case impossible at compile time, see `TODO` above mentioning `never` which is equivalent.

**Comment by mxinden** [comment]
> Note for self: [Higher-Rank Trait Bounds - The Rustonomicon](https://doc.rust-lang.org/nomicon/hrtb.html)
> 
> I learned a lot, thank you!

To be honest, this is the first time I had to make use of it. And I didn't quite understand its purpose before. The Rust type system is very powerful. Especially once we have https://rust-lang.github.io/rfcs/1598-generic_associated_types.html. It is a very exciting language.

**Comment by ackintosh** [comment]
Yeah, Rust is very exciting. I started programming with dynamically typed languages (e.g. PHP, Ruby), now I love Rust. Rust gives me a sense of _designing the types_.

**Comment by ackintosh** [comment]
> I think the actual benchmark comparison would be more fruitful. E.g. running on both versions and then comparing via https://github.com/BurntSushi/critcmp.

I'll do benchmark comparison, thanks for letting me know it. ✨

**Pull request merged**